### PR TITLE
Sagemaker training and inference integration

### PR DESCRIPTION
**Pull request by shamika**
This pull request includes :

1. How to train a model using sagemkaer 
2. How to deploy custom trained model and a model directly from sagemaker as a sagemaker realtime endpoint

### CODE REVIEW

File: framework-integrations/sagemaker/training/code/conf/config.yaml [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# The Defaults List is a list in an input config that instructs Hydra how to build the output config.
+defaults:
+  - _self_
+  - dataset: uniref50
```
Comment (qfournier) [inline framework-integrations/sagemaker/training/code/conf/config.yaml:4]: `uniref100+.yaml` should be the default.

File: framework-integrations/sagemaker/training/code/conf/dataset/uniref100.yaml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+train:
+  samples_before_next_set: null
+  paths: {uniref100: uniref/mila/uniref100_train.csv}
```
Comment (qfournier) [inline framework-integrations/sagemaker/training/code/conf/dataset/uniref100.yaml:3]: Why is `uniref50: /opt/ml/input/data/train/uniref50_sample_100.csv` but `uniref100: uniref/mila/uniref100_train.csv`?

**Comment by bschul02-amgen** [comment]
Thanks for putting this up!

I tried running `framework-integrations/sagemaker/training/sagemaker-training.ipynb` and got this error at the code cell under the heading "Start Training":

```
{
	"name": "ValueError",
	"message": "No file named \"train.py\" was found in directory \"code\".",
	"stack": "---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[14], line 4
      1 training_job_name = f\"AMPLIFY-hf-training-job-{int(time.time())}\"
----> 4 estimator.fit({
      5     'train': s3_data_location
      6 }, job_name=training_job_name)

File ~/work/AMPLIFY-AWS/env/lib/python3.10/site-packages/sagemaker/workflow/pipeline_context.py:346, in runnable_by_pipeline.<locals>.wrapper(*args, **kwargs)
    342         return context
    344     return _StepArguments(retrieve_caller_name(self_instance), run_func, *args, **kwargs)
--> 346 return run_func(*args, **kwargs)

File ~/work/AMPLIFY-AWS/env/lib/python3.10/site-packages/sagemaker/estimator.py:1366, in EstimatorBase.fit(self, inputs, wait, logs, job_name, experiment_config)
   1301 @runnable_by_pipeline
   1302 def fit(
   1303     self,
   (...)
   1308     experiment_config: Optional[Dict[str, str]] = None,
   1309 ):
   1310     \"\"\"Train a model using the input training dataset.
   1311 
   1312     The API calls the Amazon SageMaker CreateTrainingJob API to start
   (...)
   1364         :class:`~sagemaker.workflow.pipeline_context.PipelineSession`
   1365     \"\"\"
-> 1366     self._prepare_for_training(job_name=job_name)
   1368     experiment_config = check_and_get_run_experiment_config(experiment_config)
   1369     self.latest_training_job = _TrainingJob.start_new(self, inputs, experiment_config)

File ~/work/AMPLIFY-AWS/env/lib/python3.10/site-packages/sagemaker/estimator.py:3589, in Framework._prepare_for_training(self, job_name)
   3581 def _prepare_for_training(self, job_name=None):
   3582     \"\"\"Set hyperparameters needed for training. This method will also validate ``source_dir``.
   3583 
   3584     Args:
   (...)
   3587             constructor if applicable.
   3588     \"\"\"
-> 3589     super(Framework, self)._prepare_for_training(job_name=job_name)
   3591     self._validate_and_set_debugger_configs()

File ~/work/AMPLIFY-AWS/env/lib/python3.10/site-packages/sagemaker/estimator.py:944, in EstimatorBase._prepare_for_training(self, job_name)
    935 if self.source_dir or self.entry_point or self.dependencies:
    936     # validate source dir will raise a ValueError if there is something wrong with
    937     # the source directory. We are intentionally not handling it because this is a
    938     # critical error.
    939     if (
    940         self.source_dir
    941         and not is_pipeline_variable(self.source_dir)
    942         and not self.source_dir.lower().startswith(\"s3://\")
    943     ):
--> 944         validate_source_dir(self.entry_point, self.source_dir)
    946     # if we are in local mode with local_code=True. We want the container to just
    947     # mount the source dir instead of uploading to S3.
    948     local_code = get_config_value(\"local.local_code\", self.sagemaker_session.config)

File ~/work/AMPLIFY-AWS/env/lib/python3.10/site-packages/sagemaker/fw_utils.py:199, in validate_source_dir(script, directory)
    197 if directory:
    198     if not os.path.isfile(os.path.join(directory, script)):
--> 199         raise ValueError(
    200             'No file named \"{}\" was found in directory \"{}\".'.format(script, directory)
    201         )
    203 return True

ValueError: No file named \"train.py\" was found in directory \"code\"."
}
```

We should probably resolve this before merging.

File: framework-integrations/sagemaker/img/training.png
[the remark this answers is not in the document]
Comment (shamika) [reply framework-integrations/sagemaker/img/training.png:1]: Changed the training image to use Uniref100

File: framework-integrations/sagemaker/inference/code/inference_sm_trained.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+            print(f"File: {os.path.join(root, name)}")
+    
+    print(f"Listing files in the model directory: {model_dir}")
+    for root, dirs, files in os.walk(model_dir):
+        for name in dirs:
+            print(f"Directory: {os.path.join(root, name)}")
+        for name in files:
+            print(f"File: {os.path.join(root, name)}")
+
+    amplify_model_path = "/opt/ml/model/AMPLIFY_120M/pytorch_model.pt"
```
[the remark this answers is not in the document]
Comment (shamika) [reply framework-integrations/sagemaker/inference/code/inference_sm_trained.py]: Corrected to use parameterized model. Now you can pass the paramter as the env variable

File: framework-integrations/sagemaker/training/code/conf/config.yaml (hunk 1 above)
Comment (shamika) [reply framework-integrations/sagemaker/training/code/conf/config.yaml:4]: For sagemaker training we had used the uniref50 as sample dataset. You can change this to any training dataset.

File: framework-integrations/sagemaker/training/code/conf/dataset/uniref100.yaml (hunk 2 above)
Comment (shamika) [reply framework-integrations/sagemaker/training/code/conf/dataset/uniref100.yaml:3]: Corrected to have only the uniref50 for sagemaker. Again you can change this to any data sources while training and can provide as a parameter. Please let me know if you need to have this default to uniref100

**Comment by shamika** [comment]
@bschul02-amgen Did you download the entire pull request ? The error says that the train.py not exists but it is there on the repo. Can you let me know if there is any issue after downloading entire repo.

File: framework-integrations/sagemaker/inference/code/requirements.txt [hunk 4]
```diff
@@ -0,0 +1,2 @@
+transformers==4.37.0
+xformers
```
[the remark this answers is not in the document]
Comment (shamika) [reply framework-integrations/sagemaker/inference/code/requirements.txt]: Corrected

**Comment by shamika** [comment]
@bschul02-amgen @qfournier Any update on this pull request ? Please let me know if you have any questions with this regard. Thanks

**Comment by bschul02-amgen** [comment]
@shamika I found the cause of the issue above:

Relative paths are sensitive in some cases to the location of the Jupyter kernel, and these seem to be exactly the points at the code failed for me. (E.g.: the value of `source_dir` in the call to `PyTorchModel` in `sagemaker_inference_hf.ipynb`)

This is a largely unaddressed issue with paths in a Jupyter environment -- not intrinsic to your code. Even so, I'd like for the notebook to work "out of the box" across as many possible scenarios as possible, so that it's easy for new users to get up and running with it.

As such, my last ask (since everything else looks fine) is this: could we add a note somewhere in the notebook itself or any relevant documentation that makes it clear what Jupyter's working directory needs to be in order for the notebook paths to resolve? Or, failing that, what should be the intended invocation of the notebook or associated Jupyter environment?

We really appreciate your work on this, and I really hope this doesn't seem like nitpicking. If you don't mind making this last change, though, I would be happy to merge this.

**Comment by shamika** [comment]
> @shamika I found the cause of the issue above:
> 
> Relative paths are sensitive in some cases to the location of the Jupyter kernel, and these seem to be exactly the points at the code failed for me. (E.g.: the value of `source_dir` in the call to `PyTorchModel` in `sagemaker_inference_hf.ipynb`)
> 
> This is a largely unaddressed issue with paths in a Jupyter environment -- not intrinsic to your code. Even so, I'd like for the notebook to work "out of the box" across as many possible scenarios as possible, so that it's easy for new users to get up and running with it.
> 
> As such, my last ask (since everything else looks fine) is this: could we add a note somewhere in the notebook itself or any relevant documentation that makes it clear what Jupyter's working directory needs to be in order for the notebook paths to resolve? Or, failing that, what should be the intended invocation of the notebook or associated Jupyter environment?
> 
> We really appreciate your work on this, and I really hope this doesn't seem like nitpicking. If you don't mind making this last change, though, I would be happy to merge this.

Thank you for your thoughtful feedback! This is a great point and definitely not nitpicking—ensuring that the notebook works "out of the box" across different setups is crucial, especially for new users.

To address this, I've added a note in the notebook that clarifies what the working directory should be in order for the relative paths to resolve correctly. Specifically, the notebook now includes instructions explaining how to check and set the working directory to <repo-root>/framework-integrations/sagemaker/training. I also provided guidance on how to launch Jupyter from the correct directory to avoid path-related issues, along with code snippets to either verify or manually adjust the working directory.

**Comment by bschul02-amgen** [comment]
Fantastic -- thank you!

Will merge.

**Pull request merged**